File: typescript/src/templateEngine/blocks/block.ts
// Copyright (c) Microsoft. All rights reserved.

import { ILogger, NullLogger } from '../../utils/logger';

/**
 * Base class for blocks parsed from a prompt template.
 */
export abstract class Block {
    // The block content.
    private readonly _content: string;

    // Application logger.
    protected readonly log: ILogger;

    /**
     * Base constructor.
     *
     * @constructor
     * @param content Block content.
     * @param log Application logger.
     */
    protected constructor(content?: string, log?: ILogger) {
        this.log = log ?? new NullLogger();
        this._content = content ?? '';
    }

    /**
     * Check if the block content is valid.
     *
     * @returns True if the block content is valid.
     */
    public abstract isValid(): boolean;
}

File: dotnet/src/SemanticKernel.Abstractions/TemplateEngine/Blocks/Block.cs
// Copyright (c) Microsoft. All rights reserved.

using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Microsoft.SemanticKernel.TemplateEngine.Blocks;

/// <summary>
/// Base class for blocks parsed from a prompt template
/// </summary>
public abstract class Block
{
    internal virtual BlockTypes Type => BlockTypes.Undefined;

    // internal virtual bool? SynchronousRendering => null;

    /// <summary>
    /// The block content
    /// </summary>
    internal string Content { get; }

    /// <summary>
    /// App logger
    /// </summary>
    protected ILogger Logger { get; } = NullLogger.Instance;

    /// <summary>
    /// Base constructor
    /// </summary>
    /// <param name="content">Block content</param>
    /// <param name="logger">App logger</param>
    protected Block(string? content, ILogger? logger = null)
    {
        if (logger != null) { this.Logger = logger; }

        this.Content = content ?? string.Empty;
    }

    /// <summary>
    /// Check if the block content is valid.
    /// </summary>
    /// <param name="errorMsg">Error message in case the content is not valid</param>
    /// <returns>True if the block content is valid</returns>
    public abstract bool IsValid(out string errorMsg);
}
